### DISCUSSION TITLE
Doubt: How is the dockerfile able to generate executable as there is no command that builds it

### DISCUSSION THREAD

**Discussion by AJ-54** [open-telemetry/opentelemetry-collector#3459] [Q&A] [answered]
I am not able to understand how the [Dockerfile](https://github.com/open-telemetry/opentelemetry-collector/blob/main/cmd/otelcol/Dockerfile) is able to run the command `ENTRYPOINT["/otelcol"]` as there is no executable. how does the dockerfile instruct to run the `main.go` file and create an executable out of it?

**Comment by mx-psi** [comment, marked as answer]
The Makefile task first builds the binary and copies it to the `cmd/otelcol` folder https://github.com/open-telemetry/opentelemetry-collector/blob/3f5c7180c51ed67a6f54158ede5e523822e9659e/Makefile#L175-L176 then the Dockerfile copies the binary from the build context folder to the `otelcol` image https://github.com/open-telemetry/opentelemetry-collector/blob/3f5c7180c51ed67a6f54158ede5e523822e9659e/cmd/otelcol/Dockerfile#L5 and finally it gets copied to the last stage https://github.com/open-telemetry/opentelemetry-collector/blob/3f5c7180c51ed67a6f54158ede5e523822e9659e/cmd/otelcol/Dockerfile#L16 (this is a [multi-stage build](https://docs.docker.com/develop/develop-images/multistage-build/))

**Reply by AJ-54** [reply]
Thank you so much @mx-psi !!!
